Dedicated filename input and default queue method

CasperSchippers opened the pull request:
Something I've been wanting to add for quite a while now (initial idea based on #317 and #423).

This PR adds
- a dedicated filename input field (similar to the directory input field that has been added in #317).
- The input field also has a toggle which allows to easily switch between storing the measurement in a file, or to a temporary file.
- I propose to add a default queue method: together with the directory input we have everything we need for a default queue method, reducing the boilerplate code needed to get a working GUI. The queue method doesn't work if the filename and directory input fields are disabled (which can be done via keyword arguments in the ManagedWindows). I hence also removed the queue methods in the examples (whenever applicable).
- I propose to enable the filename and directory input-widgets by default.

I'd like to hear what you think of this?

bmoneke commented:
Good initiative, sounds good.

bmoneke commented on pymeasure/display/windows/managed_window.py:
```diff
@@ -155,6 +163,15 @@ def __init__(self,
 
     def _setup_ui(self):
+        if self.filename_input:
+            self.writefile_toggle = QtWidgets.QCheckBox('Write file', self)
+            self.writefile_toggle.setLayoutDirection(QtCore.Qt.RightToLeft)
+            self.writefile_toggle.setChecked(True)
+            self.writefile_toggle.stateChanged.connect(self.toggle_file_dir_input_active)
```
I like tooltips, which describe, what an option does. Maybe you could add one.
I guess this toggle switches between writing a file into the directory and a temporary file, doesn`t it?

bmoneke commented on pymeasure/display/windows/managed_window.py:
```diff
@@ -107,8 +110,11 @@ class ManagedWindowBase(QtWidgets.QMainWindow):
         code:`Load sequence` button
     :param inputs_in_scrollarea: boolean that display or hide a scrollbar to the input area
     :param directory_input: specify, if present, where the experiment's result will be saved.
+    :param filename_input: specify, if present, the base of the filename where the results will be
+        saved.
```
How will the full name of the filename look like?

bmoneke commented on pymeasure/display/widgets/filename_widget.py:
```diff
@@ -0,0 +1,67 @@
+
+    def __init__(self, parent=None):
+        super().__init__(parent=parent)
+
```
The content of this widget does not describe the full filename, does it?
It is just the base, which will be made unique and will be expanded by the file type extension, isn't it?

I'd add a tooltip, which states, that this is the base of the filename.

bmoneke commented on pymeasure/experiment/results.py:
```diff
@@ -69,8 +69,9 @@ def replace_placeholders(string, procedure, date_format="%Y-%m-%d", time_format=
-    parameters = procedure.parameter_objects()
+    parameters = procedure.placeholder_objects()
     placeholders = {param.name: param.value for param in parameters.values()}
+    placeholders.update({varname: param.value for varname, param in parameters.items()})
```
does not placeholder_objects already contain the parameters, such that line 73 is superfluous?

CasperSchippers replied:
You are absolutely right; I had removed that but missed it in a commit.

As some context, I was hoping to use some of those nice 'in-string' calculations, like you can do with f-strings (`f"{delay * 10:.2f}"` for example, such that you can format you units nicely for example); for this i would need to add the variable-names of a parameter as a valid placeholder (since the names containing spaces won't work). However, I quickly came to realize that this wont work at all when using the `format` method, rather than f-strings.... unfortunately.

bmoneke reviewed (approved):
Generally good suggestion. I left a few questions/suggestions, especially regarding documentation (I like tooltips to explain more, what a button/checkbox etc. does).

CasperSchippers commented:
I've restructured some things a bit: I moved the directory and filename input to a new widget (FileInputWidget) and moved the scroll-area optionality to the InputsWidget (and removed some margins)

CasperSchippers commented on pymeasure/display/widgets/filename_widget.py:
```diff
@@ -0,0 +1,152 @@
+        self.setToolTip(
+            "The filename of the file to which the measurement will be stored. Placeholders (in \n"
+            "standard python format, i.e.: '{variable name:formatspec}') will be replaced by \n"
+            "the respective value. The extension '.csv' will be appended, unless an extension\n"
+            "(one of '.txt', or '.csv') is recognized. Additionally, an index number ('_#') is\n"
+            "added to ensure the uniqueness of the filename.\n"
```
Note to self: do not forget to implement stripping the extension.

CasperSchippers commented on pymeasure/display/widgets/filename_widget.py:
```diff
@@ -0,0 +1,152 @@
+            "The filename of the file to which the measurement will be stored. Placeholders (in \n"
+            "standard python format, i.e.: '{variable name:formatspec}') will be replaced by \n"
+            "the respective value. The extension '.csv' will be appended, unless an extension\n"
+            "(one of '.txt', or '.csv') is recognized. Additionally, an index number ('_#') is\n"
```
Note to self: add attributes for default and recognized extensions.

CasperSchippers commented on docs/tutorial/graphical.rst:
```diff
@@ -408,16 +378,15 @@ An example of such a sequence file is given below, resulting in the sequence sho
 
-Using the directory input
-~~~~~~~~~~~~~~~~~~~~~~~~~
+The filename and directory input
```
Note to self: modify this section.

CasperSchippers commented:
I think it's ready now.

@bmoneke, since I did a bit of reworking (e.g. moving the filename and directory input fields to a seperate widget, and some small visual changes to the inputswidget), could you have another look (can be quick I think)?

bmoneke commented on docs/tutorial/graphical.rst:
```diff
@@ -208,6 +203,65 @@ Now that you have learned about the ManagedWindow, you have all of the basics to
+    :alt: The filename and directory input widget showing the auto-complete for the directory
+    :width: 24%
+
+With the filename input, additionally a toggle (named :guilabel:`Write file` controlling whether the measurement is save at all, or should be stored in a temporary file, is displayed.
+When the toggle is ticked, the measurement is stored to a temporary file, and the other input fields are disabled.
```
A more concise description
```suggestion
If the checkbox named :guilabel:`Write file` is enabled, the measurement is written to a file.
Otherwise, it is stored in a temporary file.
```

From the code below, I noticed an inconsistency:
You **un**check "write file" in order to write to a file?
For me it is clearer to tick "write file" to write it actually.
Alternatively, it could be labelled "temp file".

CasperSchippers replied:
Ah yeah, that is in inconsistency in the documentation; it is exactly as you say: you write to file when 'write file' is ticked

bmoneke commented on docs/tutorial/graphical.rst:
```diff
@@ -208,6 +203,65 @@ Now that you have learned about the ManagedWindow, you have all of the basics to
+With the filename input, additionally a toggle (named :guilabel:`Write file` controlling whether the measurement is save at all, or should be stored in a temporary file, is displayed.
+When the toggle is ticked, the measurement is stored to a temporary file, and the other input fields are disabled.
+
+When the measurement is to be stored in to a certain file, ensure that the :guilabel:`Write file` is disabled.
```
It doesn't seem right to me to check "write file" in order to **not** write to a file!

CasperSchippers replied:
Yes, you are right, this should reald "enabled". With your clarified sentence above, this one is unnecessary.

bmoneke commented on docs/tutorial/graphical.rst:
```diff
@@ -208,6 +203,65 @@ Now that you have learned about the ManagedWindow, you have all of the basics to
+When the toggle is ticked, the measurement is stored to a temporary file, and the other input fields are disabled.
+
+When the measurement is to be stored in to a certain file, ensure that the :guilabel:`Write file` is disabled.
+The filename in the designated field can be entered with or without extension; if the entered extension is recognized (by default :code:`.csv` and :code:`.txt`) are recognized, that extension is used.
```
```suggestion
The filename in the designated field can be entered with or without extension.
If the entered extension is recognized (by default :code:`.csv` and :code:`.txt` are recognized), that extension is used.
```

Parenthesis was misleading.

bmoneke commented on docs/tutorial/graphical.rst:
```diff
@@ -208,6 +203,65 @@ Now that you have learned about the ManagedWindow, you have all of the basics to
+
+
+The fields for the filename and the directory can be disabled using the boolean keyword arguments :code:`filename_input` and :code:`directory_input`, respectively, in the :class:`~pymeasure.display.windows.managed_window.ManagedWindow` init.
+Note that when a either of these values is set to :code:`False`, the default :meth:`~pymeasure.display.windows.managed_window.ManagedWindowBase.queue` method of the :class:`~pymeasure.display.windows.managed_window.ManagedWindow` class will no longer work, and a new, custom, method needs to be implemented; a basic implementation is shown in the documentation of the :meth:`~pymeasure.display.windo ...
```
```suggestion
Note that when either of these values is set to :code:`False`, the default :meth:`~pymeasure.display.windows.managed_window.ManagedWindowBase.queue` method of the :class:`~pymeasure.display.windows.managed_window.ManagedWindow` class will no longer work, and a new, custom, method needs to be implemented; a basic implementation is shown in the documentation of the :meth:`~pymeasure.display.windows.managed_window.ManagedWindowBase.queue` method.
```

bmoneke reviewed (approved):
A few suggestions on wording and a suggestion of coding.

CasperSchippers commented on pymeasure/display/widgets/fileinput_widget.py:
```diff
@@ -0,0 +1,153 @@
+class FileInputWidget(QtWidgets.QWidget):
+    _extensions = ["csv", "txt"]
+
+    def __init__(self, filename_input=True, directory_input=True, parent=None):
```
You are right it is a bit of an uggly construction right now. The MixIn classes is an option, but I think you are right that is is more complicated.

When rethinking this, I _think_ that it is not really necessary to be able to select only the filename or directory input field. So now I went for:
- Just a single class, no option to remove one of the fields. So also only a single boolean toggle in the main-window `enable_file_input` instead of the two filename and directory toggles.
- I did add a flag `filename_fixed` that allows for 'freezing' the filename. In that case, the filename field will be disabled, but still present; such that you clearly see where the file will be saved, but are not able to change it.

What do you think?

bmoneke replied:
I like it.

CasperSchippers commented on pymeasure/display/widgets/fileinput_widget.py:
```diff
@@ -0,0 +1,153 @@
+    def __init__(self, filename_input=True, directory_input=True, parent=None):
+        super().__init__(parent)
+
+        self._setup_ui(filename_input, directory_input)
+        self._layout()
```
Yeah, I understand your line of thought, especially with the two `hasattr` checks that were in there.
The idea of this separation is that you define the widgets (that you need/can access later on in the class) in the `setup_ui` method, and have the `layout` method (which can sometimes be quite full and cluttered) to handle where the widgets are placed.

SInce I've removed the `hasattr` check, i'm inclined to keep the structure like this.

bmoneke replied:
> SInce I've removed the hasattr check, i'm inclined to keep the structure like this.

It's fine with me to keep the structure, which makes sense to me.

Only the "hasattr" checks made it seem better to unite the methods.

CasperSchippers commented on pymeasure/display/windows/managed_window.py:
```diff
@@ -123,7 +134,8 @@ def __init__(self,
                  sequence_file=None,
                  inputs_in_scrollarea=False,
-                 directory_input=False,
+                 directory_input=True,
+                 filename_input=True,
```
Yeah, I agree. changed.

CasperSchippers commented:
Thank you, @bmoneke, for the reviewing and thinking along!

bmoneke commented:
You're welcome. It show, that review are a great tool for good code (I notice it with pyleco a lot as well).
